e2e_test/js-tests/test_viem_smoketest.mjs

```diff
@@ -0,0 +1,134 @@
+import { assert } from "chai";
+import "mocha";
+import {
+	parseAbi,
+} from "viem";
+import fs from "fs";
+import { publicClient, walletClient } from "./viem_setup.mjs"
+
+// Load compiled contract
+const testContractJSON = JSON.parse(fs.readFileSync(process.env.COMPILED_TEST_CONTRACT, 'utf8'));
+
+// check checks that the receipt has status success and that the transaction
+// type matches the expected type, since viem sometimes mangles the type when
+// building txs.
+async function check(txHash, type) {
+	const receipt = await publicClient.waitForTransactionReceipt({ hash: txHash });
+	assert.equal(receipt.status, "success", "receipt status 'failure'");
+	const transaction = await publicClient.getTransaction({ hash: txHash });
+	assert.equal(transaction.type, type, "transaction type does not match");
+}
+
+// sendTypedTransaction sends a transaction with the given type and an optional
+// feeCurrency.
+async function sendTypedTransaction(type, feeCurrency) {
+	return await walletClient.sendTransaction({
+		to: "0x00000000000000000000000000000000DeaDBeef",
+		value: 1,
+		type: type,
```

Comment:
viem doesnt use this type field.

Reply:
I think it must be using it because the tests verify that the transaction retrieved from the node has the same type passed to sendTransaction. How else could the types be matching up, see [here](https://github.com/celo-org/op-geth/pull/261/files#diff-729cf14983159ee568c6e3d85cc7d4b92cb7c7271fc160fe79dc7dfc4753e915R19)

Reply:
so kinda the opposite. 

when you deserialize / recover the transaction from the rlp it adds the type. 


I was slightly in correct it does seem like viem will use the field to serialize if the type is there and known but if its not known it uses heuristics like does the transaction have gasPrice or does it have maxFeePerGas field to figure out type.

Reply:
Ok got you, thanks @aaronmgdr!